### PR TITLE
Added jupyter notebook for NE double loop

### PR DESCRIPTION
**Pull request by radhakrishnatg**
## Addresses issue:
#81 

## Changes proposed in this PR:
- Adds a Jupyter notebook to demonstrate the double loop workflow for the generic nuclear case study

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.md and COPYRIGHT.md file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by lbianchi-lbl** [comment]
> Thank you @lbianchi-lbl ! We do specify output directory as one of Prescient options:
> 
> ```python
> options = {
>     "data_path": rts_gmlc_data_dir,
>     "input_format": "rts-gmlc",
>     "simulate_out_of_sample": True,
>     "run_sced_with_persistent_forecast_errors": True,
>     "output_directory": "bidding_plugin_test_multiperiod_nulcear",...}
> ```
> 
> Also, the notebook I'm using is the one in #103 not the one in this PR

@radhakrishnatg Thanks for the clarification, this helps. I suggest to move the conversation from #119 to this PR since I think the latest error is more directly connected with the notebook here.

---


```---------------------------------------------------------------------------
PermissionError                           Traceback (most recent call last)
<ipython-input-8-3078a83579e4> in <module>
     32 
     33 # Run Prescient
---> 34 Prescient().simulate(**options)

c:\users\radhakrishna\desktop\prescient\prescient\simulator\prescient.py in simulate(self, **options)
     73             self.config.set_value(options)
     74 
---> 75         return self._simulate(self.config)
     76 
     77     def _simulate(self, options: PrescientConfig):

c:\users\radhakrishna\desktop\prescient\prescient\simulator\prescient.py in _simulate(self, options)
     80                                 "If you wish to simulate again create a new Prescient object.")
     81         self.simulate_called = True
---> 82         return super().simulate(options)
     83 
     84 def main(args=None):

c:\users\radhakrishna\desktop\prescient\prescient\simulator\simulator.py in simulate(self, options)
     88         oracle_manager.initialize(engine, data_manager, options)
     89         stats_manager.initialize(options)
---> 90         reporting_manager.initialize(options, stats_manager)
...
--> 398                 os.unlink(fullname)
    399             except OSError:
    400                 onerror(os.unlink, fullname, sys.exc_info())

PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'c:\\Users\\Radhakrishna\\Desktop\\DISPATCHES\\dispatches\\models\\nuclear_case\\flowsheets\\bidding_plugin_test_multiperiod_nulcear\\bus_detail.csv'
```

Based on the error message above, I think this might be unrelated from the input data files (i.e. the `dispatches-sample-data` package that we've been debugging over the past few days). Is it possible that the file mentioned in the error message `bus_detail.csv` is being used in the same notebook? I remember that in a separate conversation you asked me how to access files in the `bidding_plugin_test_multiperiod_nuclear` directory, which might or might not be related to the error.

If (as I seem to undestand from the notebook) the analysis (i.e. the loading of the files from `bidding_plugin_test_multiperiod_nuclear`) happens after the simulation is initiated (which causes the deletion of the `bidding_plugin_test_multiperiod_nuclear` directory), then the notebook will work only if you run the cells in order. You can use the "Restart kernel and run all" menu for this:

image

(I'm aware this is a different notebook from the one in this PR, but this is just an example since the menu will look the same).

**Comment by radhakrishnatg** [comment]
Thank you @lbianchi-lbl ! I'm sorry for the confusion! There are two issues that I need to resolve:
(1) Run Prescient with the RTS-GMLC dataset located in `dispatches-sample-data` package.
(2) Make plots of the results after Prescient simulation is done.

The error message on `PermissionError` is related to the first one. I got the error when I ran Prescient with `dispatches-sample-data` package. You are right. I should have restarted the kernel. When I did that I didn't receive any permission error. So I think this issue is resolved. 

Loading the files from `bidding_plugin_test_multiperiod_nuclear` relates to the second issue. After Prescient finishes the simulation, it will create a new folder titled `bidding_plugin_test_multiperiod_nuclear` in the same folder where my Jupyter notebook is present, and it contains the results I need. For some reason, I was not able to read the csv file using `pd.read_csv("bidding_plugin_test_multiperiod_nuclear/tracker_detail.csv")`. I do not understand why this is happening. You pointed me to a wiki page the other day. I will now try that and see if it works. Thank you for your help, @lbianchi-lbl !

**Comment by xiangao1** [comment]
What's the error message?

**Comment by radhakrishnatg** [comment]
@xiangao1 The default P_min and P_max were 400 MW and 396 MW. For the NE case study, I'm using P_min = 400 and P_max = 500. I think it took the P_min value to be 396 MW somewhere, so it threw a value error. 

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-e1cf36381538> in <module>
     33 
     34 # Run Prescient
---> 35 Prescient().simulate(**options)

c:\users\radhakrishna\desktop\prescient\prescient\simulator\prescient.py in simulate(self, **options)
     73             self.config.set_value(options)
     74 
---> 75         return self._simulate(self.config)
     76 
     77     def _simulate(self, options: PrescientConfig):

c:\users\radhakrishna\desktop\prescient\prescient\simulator\prescient.py in _simulate(self, options)
     80                                 "If you wish to simulate again create a new Prescient object.")
     81         self.simulate_called = True
---> 82         return super().simulate(options)
     83 
     84 def main(args=None):

c:\users\radhakrishna\desktop\prescient\prescient\simulator\simulator.py in simulate(self, options)
     93 
     94         first_time_step = time_manager.get_first_time_step()
---> 95         oracle_manager.call_initialization_oracle(options, first_time_step)
...
    726 
    727 

ValueError: Invalid parameter value: RegulationLowLimit[('121_NUCLEAR_1', 1)] = '3.96', value type=<class 'float'>.
	Value failed parameter validation rule
```

**Comment by xiangao1** [comment]
@bknueven Is there something special we need to do for nuclear?

**Comment by bknueven** [comment]
> @bknueven Is there something special we need to do for nuclear when passing the static params?

If you're using the full RTS-GMLC stack, then you should also set the `p_min_agc` and `p_max_agc` parameters to `p_min` and `p_max`.

However, I would recommend instead just deleting this file: https://github.com/gmlc-dispatches/sample-data/blob/9a04c18bf8b0d5a74b497e77df61eab4e542775e/dispatches_sample_data/rts_gmlc/SourceData/reserves.csv. That will remove the AGC requirement for the system, and run RTS-GMLC in a fashion more similar to using the old *.dat files for Prescient. This is what I have done for the parameter sweeps. @lbianchi-lbl can make the appropriate change, if we think it's a good idea.

**Comment by lbianchi-lbl** [comment]
> > @bknueven Is there something special we need to do for nuclear when passing the static params?
> 
> If you're using the full RTS-GMLC stack, then you should also set the `p_min_agc` and `p_max_agc` parameters to `p_min` and `p_max`.
> 
> However, I would recommend instead just deleting this file: https://github.com/gmlc-dispatches/sample-data/blob/9a04c18bf8b0d5a74b497e77df61eab4e542775e/dispatches_sample_data/rts_gmlc/SourceData/reserves.csv. That will remove the AGC requirement for the system, and run RTS-GMLC in a fashion more similar to using the old *.dat files for Prescient. This is what I have done for the parameter sweeps. @lbianchi-lbl can make the appropriate change, if we think it's a good idea.

Thanks for the suggestion @bknueven. Just one clarification: if I delete `SourceData/reserves.csv` from the `dispatches-sample-data` package, would it still be compatible with the double-loop workflows for the other case studies? If yes, then I can do this immediately and upload a new version of `dispatches-sample-data` on PyPI.

**Comment by lbianchi-lbl** [comment]
OK, I've just released a new version without the `SourceData/reserves.csv` file: https://pypi.org/project/dispatches-sample-data/22.6.28/.

@radhakrishnatg can you try to install and test the new version? Make sure that new version is being used in the notebook by e.g. using "Restart kernel and run all" for the Jupyter notebook.

```sh
pip install -i dispatches-sample-data==22.06.28
python -c "from dispatches_sample_data import __version__ as v; print(v)"
python -c "from dispatches_sample_data import rts_gmlc as data; [print(p) for p in sorted(data.path.rglob('*.csv'))]"
```

**Comment by radhakrishnatg** [comment]
Thank you @lbianchi-lbl ! I installed the new version, and the tests ran without any issues.

Thank you @xiangao1 @bknueven ! I ran the double loop with the new data set. It printed a message at the beginning:
```
Initializing simulation...
Did not find reserves.csv; assuming no reserves
```
But I'm still getting the same error message:
```
ERROR: Constructing component 'RegulationLowLimit' from data=None failed:
    ValueError: Invalid parameter value: RegulationLowLimit[('121_NUCLEAR_1',
    1)] = '3.96', value type=<class 'float'>.
    	Value failed parameter validation rule
```

**Comment by bknueven** [comment]
> But I'm still getting the same error message:
> 
> ```
> ERROR: Constructing component 'RegulationLowLimit' from data=None failed:
>     ValueError: Invalid parameter value: RegulationLowLimit[('121_NUCLEAR_1',
>     1)] = '3.96', value type=<class 'float'>.
>     	Value failed parameter validation rule
> ```

Well, I'm not sure why this code is still getting called. Regardless, removing the reserves will result in a faster simulation

@xiangao1 we should set the `p_min_agc` and `p_max_agc` attributes on the generator to `p_min` and `p_max` somewhere in the double-loop code, just to guard against this case.

**Comment by xiangao1** [comment]
Just created a PR in IDAES: https://github.com/IDAES/idaes-pse/pull/899. @radhakrishnatg Please have a try.

**Comment by lbianchi-lbl** [comment]
@xiangao1 unfortunately having to incorporate unreleased code from IDAES isn't really feasible at this point of the DISPATCHES release cycle.

Since the changes seem relatively minor, would it be possible to define a subclass with those changes inside the DISPATCHES codebase? e.g. create a `dispatches/grid_integration_utils.py` file and define something like this

```py
from idaes.apps.grid_integration.model_data import (
    GeneratorModelData as _UpstreamGeneratorModelData,
    RealValueValidator,
    AtLeastPminValidator
)


class GeneratorModelData(_UpstreamGeneratorModelData):
    p_min_agc = RealValueValidator(min_val=0)
    p_max_agc = AtLeastPminValidator()

    def __init__(
        self,
        gen_name,
        generator_type,
        p_min,
        p_max, *args, **kwargs):
        self.p_min_agc = p_min
        self.p_max_agc = p_max
        super().__init__(*args, gen_name, generator_type, p_min=p_min, p_max=p_max, **kwargs)
```

**Comment by xiangao1** [comment]
@lbianchi-lbl sure! That's something @radhakrishnatg should do in his code.

**Comment by radhakrishnatg** [comment]
Thank you @xiangao1 ! Now, I can run the double loop notebook without any issues after including the changes in your PR.

@lbianchi-lbl I want your suggestion regarding this. If I'm not wrong the notebook will not be tested while merging the PR with DISPATCHES main. So, I'm assuming that it will not lead to any test failures. The only drawback is that the user needs to do some manual changes in order to be able to run the notebook. So, is it better to provide instructions in the notebook on what changes are needed to be made or add a file to the dispatches repo with the class definition mentioned in your comment?

**Comment by lbianchi-lbl** [comment]
> Thank you @xiangao1 ! Now, I can run the double loop notebook without any issues after including the changes in your PR.
> 
> @lbianchi-lbl I want your suggestion regarding this. If I'm not wrong the notebook will not be tested while merging the PR with DISPATCHES main. So, I'm assuming that it will not lead to any test failures. The only drawback is that the user needs to do some manual changes in order to be able to run the notebook. So, is it better to provide instructions in the notebook on what changes are needed to be made or add a file to the dispatches repo with the class definition mentioned in your comment?

@radhakrishnatg You could try to add the code from my previous comment directly in the notebook by adding a cell before the cell where the `GeneratorModelData` object is instantiated and used, and then run the notebook to verify it works. If needed, I can then go and extract that code when I set up the Jupyter notebook testing infrastructure.

**Comment by radhakrishnatg** [comment]
Thank you @lbianchi-lbl ! I added the class to the notebook. 

@dguittet @jghouse88 @nareshsusarla Could you please review the PR when you have some time? I want to add a few more results and some additional comparisons, but it will take time. So, I will do those later and create a new PR for that. Thank you!

File: dispatches/models/nuclear_case/flowsheets/nuclear_flowsheet_multiperiod_class.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+    @property
+    def power_output(self):
+        return "P_T"
+
+    @property
+    def total_cost(self):
+        return ("tot_cost", 1)
+
+    # @property
```
Comment (nareshsusarla) [inline dispatches/models/nuclear_case/flowsheets/nuclear_flowsheet_multiperiod_class.py:256]: If these will not be used in future, then may be consider removing these lines

**Review by nareshsusarla — APPROVED** [review]
looks good

File: dispatches/models/nuclear_case/flowsheets/nuclear_flowsheet_fixed_design.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        h2_price: Price of hydrogen (in $/kg)
+        demand_type: "fixed"/"variable" hydrogen demand
+
+    Returns:
+        m: Object containing the nuclear flowsheet
+    """
+    mw_h2 = 2.016 * 1e-3  # Molecular mass of hydrogen in kg/mol
+
+    # Define conversion factors
+    MW_to_kW = 1000
```
[the remark this answers is not in the document]
Comment (radhakrishnatg) [reply dispatches/models/nuclear_case/flowsheets/nuclear_flowsheet_fixed_design.py:67]: Added to TODO for now. I will make the change later.

File: dispatches/models/nuclear_case/flowsheets/nuclear_flowsheet_multiperiod_class.py (hunk 1 above)
Comment (radhakrishnatg) [reply dispatches/models/nuclear_case/flowsheets/nuclear_flowsheet_multiperiod_class.py]: Removed the lines from the file.

**Review by nareshsusarla — APPROVED** [review]

**Pull request merged**